java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow;
+
+import static org.apache.arrow.vector.types.FloatingPointPrecision.DOUBLE;
+import static org.apache.arrow.vector.types.FloatingPointPrecision.SINGLE;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.BaseFixedWidthVector;
+import org.apache.arrow.vector.BigIntVector;
+import org.apache.arrow.vector.BitVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.Float4Vector;
+import org.apache.arrow.vector.Float8Vector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.arrow.vector.VarCharVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.complex.impl.VarBinaryWriterImpl;
+import org.apache.arrow.vector.holders.BigIntHolder;
+import org.apache.arrow.vector.holders.BitHolder;
+import org.apache.arrow.vector.holders.Float4Holder;
+import org.apache.arrow.vector.holders.Float8Holder;
+import org.apache.arrow.vector.holders.IntHolder;
+import org.apache.arrow.vector.holders.VarBinaryHolder;
+import org.apache.arrow.vector.holders.VarCharHolder;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Type;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.util.Utf8;
+
+/**
+ * Class that does most of the work to convert Avro data into Arrow columnar format Vector objects.
+ */
+public class AvroToArrowUtils {
```

Comment:
I'm not familiar with the avro APIs but I think the write approach here is to implement a [DatumReader](https://github.com/apache/avro/blob/2d3b1fe7efd865639663ba785877182e7e038c45/lang/java/avro/src/main/java/org/apache/avro/io/DatumReader.java), I'm not sure if this should be an extension of GenericDatumReader or a new class either (would need to read the code for it).  I'm not sure it generic D should actually be a VectorSchemaRoot or a wrapper around it either (again would have to prototype something) but doing it at that level makes the decoding generic across any sort of transport and parallelization schemes.

I would also see if there is a more OO approach that caches individual FieldWriters to populate them directly from the decoded primitives, rather then a switch statement/static method.

Reply:
Thanks for comments. Now this class accept DataFileReader to read avro data(which already holds DatumReader), do you mean we should accept param like DatumReader<VectorSchemaRoot> to read and covert data?

Reply:
And it generic D seems should be a avro type like Record, Enum, Int etc and seems impossible to make something like \<VectorSchemaRoot\>.

>'I would also see if there is a more OO approach that caches individual FieldWriters to populate them directly from the decoded primitives, rather then a switch statement/static method.'
IMO, can you give some suggestion with current implementation? thanks

Reply:
I think the goals for this library should be:
* Make it is as fast as possible.  I think this probably comes in two rounds:
     1.  Integrate with the Avro library at the decoder level to avoid boxing/unboxing.
     2.  At some point use bytebuddy or another generator library to avoid schema interpretation (this should come after we have something working first)
* Widely applicable.  Right now the code seems file specific, it would also be nice if it could work on arbitrary avro blocks sitting in memory   .  Is seemed like creating

After a quick look it seemed to me that implementing a the DatumReader would be the best way to make it happen but if that isn't possible I would need to take a closer look.

I thought an OO approach might be cleaner because it would allow you to have a something like:
```
interface Consumer {
   
   void consume(Decoder decoder);
}
class AvroIntConsumer implements Consumer {

   IntWriter writer;
   void consume(Decoder decoder) {
         writer.writer(decoder.getInt())
   }
}
```
but its possible that this is overkill/less efficient then a switch statement.

Reply:
Thanks a lot for your valuable suggestion. I updated this PR:
1. Add consumers to read data from decoder and write to Arrow vector
2. Use GenericDatumReader(implementation of DatumReader which could handle either primitive type and complex type) instead of DataFileReader.
Could you please help take a look again? Thanks! Note now only support primitive type, still many work need to be done in follow-ups.